### PR TITLE
Add registerUpkeep as an external method

### PR DESCRIPTION
**Pull request by EasterTheBunny**

### CODE REVIEW

File: contracts/src/v0.5/dev/SchnorrSECP256K1.sol [hunk 1]
```diff
@@ -6,7 +6,8 @@ pragma solidity ^0.5.0;
 
 contract SchnorrSECP256K1 {
   // See https://en.bitcoin.it/wiki/Secp256k1 for this constant.
-  uint256 constant public Q = // Group order of secp256k1
+  // Group order of secp256k1
+  uint256 constant public Q =
```
Comment (infiloop2) [inline contracts/src/v0.5/dev/SchnorrSECP256K1.sol:10]: nit. can you remove unrelated changes?

File: contracts/src/v0.6/VRF.sol [hunk 2]
```diff
@@ -133,7 +133,8 @@ pragma solidity 0.6.6;
 contract VRF {
 
   // See https://www.secg.org/sec2-v2.pdf, section 2.4.1, for these constants.
-  uint256 constant private GROUP_ORDER = // Number of points in Secp256k1
+  // Number of points in Secp256k1
+  uint256 constant private GROUP_ORDER =
```
Comment (infiloop2) [inline contracts/src/v0.6/VRF.sol:137]: same

File: contracts/src/v0.8/dev/keeper2_0/KeeperRegistrar2_0.sol [hunk 3]
```diff
@@ -140,22 +140,35 @@ contract KeeperRegistrar2_0 is TypeAndVersionInterface, ConfirmedOwner, ERC677Re
...
+    _register(name, encryptedEmail, upkeepContract, gasLimit, adminAddress, checkData, amount, sender);
+  }
 
-      _approve(name, upkeepContract, gasLimit, adminAddress, checkData, amount, hash);
-    } else {
-      uint96 newBalance = s_pendingRequests[hash].balance + amount;
-      s_pendingRequests[hash] = PendingRequest({admin: adminAddress, balance: newBalance});
+  /**
+   * @notice Allows external users to register upkeeps
+   * @notice registerUpkeep can only be called through transferAndCall on LINK contract
```
Comment (infiloop2) [inline contracts/src/v0.8/dev/keeper2_0/KeeperRegistrar2_0.sol:148]: nit. update comment

File: contracts/test/v0.8/dev/keeper2_0/KeeperRegistrar2_0.test.ts [hunk 4]
```diff
@@ -549,6 +549,64 @@ describe('KeeperRegistrar2_0', () => {
     })
   })
 
+  describe('#registerUpkeep', () => {
```
Comment (infiloop2) [inline contracts/test/v0.8/dev/keeper2_0/KeeperRegistrar2_0.test.ts:552]: can you add another test for InsufficientPayment() error

File: contracts/test/v0.8/dev/keeper2_0/KeeperRegistrar2_0.test.ts [hunk 5]
```diff
@@ -549,6 +549,93 @@ describe('KeeperRegistrar2_0', () => {
...
+          executeGas,
+          await admin.getAddress(),
+          emptyBytes,
+          amount,
+        )
+      assert.equal((await registry.getState()).state.numUpkeeps.toNumber(), 1) // 0 -> 1
+
+      //confirm if a new upkeep has been registered and the details are the same as the one just registered
+      const [id] = await registry.getActiveUpkeepIDs(0, 1)
+      const newupkeep = await registry.getUpkeep(id)
```
Comment (infiloop2) [inline contracts/test/v0.8/dev/keeper2_0/KeeperRegistrar2_0.test.ts:627]: can you actually pick the upkeep ID from tx, instead of this logic?

Comment (infiloop2) [reply contracts/test/v0.8/dev/keeper2_0/KeeperRegistrar2_0.test.ts:627]: also another test case would be good with auto approve off and assertion that returned upkeep ID is zero

Comment (EasterTheBunny) [reply contracts/test/v0.8/dev/keeper2_0/KeeperRegistrar2_0.test.ts:627]: I might need some help with that. I tried doing so and the solution didn't come to me. I really wanted to get the returned id from the `registerUpkeep` function and compare it do the one above.

How do you extract the returned value from a `ContractTransaction`?

Comment (infiloop2) [reply contracts/test/v0.8/dev/keeper2_0/KeeperRegistrar2_0.test.ts:627]: Discussed offline, not trivial to do

**Review by infiloop2 — DISMISSED** [review]

**Review by cmalec — APPROVED** [review]
